litestar/channels/plugin.py

```diff
@@ -0,0 +1,431 @@
+
+if TYPE_CHECKING:
+    from litestar.channels.base import ChannelsBackend
+    from litestar.config.app import AppConfig
+    from litestar.connection import WebSocket
+    from litestar.types import LitestarEncodableType, TypeEncodersMap
+    from litestar.types.asgi_types import WebSocketMode
+
+
+T = TypeVar("T")
+
+
+class ChannelsException(LitestarException):
+    pass
+
+
+class AsyncDeque(Queue, Generic[T]):
+    def __init__(self, maxsize: int | None) -> None:
+        self._deque_maxlen = maxsize
+        super().__init__()
+
+    def _init(self, maxsize: int) -> None:
+        self._queue: deque[T] = deque(maxlen=self._deque_maxlen)
+
+
+BacklogStrategy = Literal["backoff", "dropleft"]
+
+
+class Subscriber:
+    def __init__(
+        self, plugin: ChannelsPlugin, max_backlog: int | None = None, backlog_strategy: BacklogStrategy = "backoff"
+    ) -> None:
+        """A wrapper around a stream of events published to subscribed channels"""
+        self._task: asyncio.Task | None = None
+        self._plugin = plugin
+        self._backend = plugin._backend
+        self._queue: Queue[bytes | None] | AsyncDeque[bytes | None]
+
+        if max_backlog and backlog_strategy == "dropleft":
+            self._queue = AsyncDeque(maxsize=max_backlog or 0)
+        else:
+            self._queue = Queue(maxsize=max_backlog or 0)
+
+    async def put(self, item: bytes | None) -> None:
+        await self._queue.put(item)
+
+    def put_nowait(self, item: bytes | None) -> bool:
+        """Put an item in the subscriber's stream without waiting"""
+        try:
+            self._queue.put_nowait(item)
+            return True
+        except QueueFull:
+            return False
+
+    @property
+    def qsize(self) -> int:
+        return self._queue.qsize()
+
+    async def iter_events(self) -> AsyncGenerator[bytes, None]:
+        """Iterate over the stream of events. If no items are available, block until
+        one becomes available
+        """
+        while True:
+            item = await self._queue.get()
+            if item is None:
+                break
+            yield item
+            self._queue.task_done()
+
+    async def put_history(self, channels: str | Sequence[str], limit: int | None = None) -> None:
+        """Fetch the history of ``channels`` from the backend and put them in the subscriber's stream"""
+        if isinstance(channels, str):
+            channels = [channels]
+
+        if len(channels) == 1:
+            await self._put_channel_history(channels[0], limit=limit)
+            return
+
+        async with anyio.create_task_group() as task_group:
+            for channel in channels:
+                task_group.start_soon(self._put_channel_history, channel, limit)
+
+    async def _put_channel_history(self, channel: str, limit: int | None = None) -> None:
+        history = await self._backend.get_history(channel, limit)
+        for entry in history:
+            self._queue.put_nowait(entry)
```

Comment:
You can do this inside a comprehension

Reply:
Sure, but using a comprehension to iterate without producing an iterable is pretty much considered an anti-pattern in my book.